[TensorExpr] Pull changes from bertmaher/pytorch_fusion.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34842 [TensorExpr] Pull changes from bertmaher/pytorch_fusion.**

This PR (hopefully the last one of such kind) is merging changes from a
side branch where tensor expessions based fuser work has been done so
far. This PR is is a squashed version of changes in the side branch,
which is available here: https://github.com/bertmaher/pytorch

Differential Revision: [D20478208](https://our.internmc.facebook.com/intern/diff/D20478208)

Commit:
[TensorExpr] Pull changes from bertmaher/pytorch_fusion.

This PR (hopefully the last one of such kind) is merging changes from a
side branch where tensor expessions based fuser work has been done so
far. This PR is is a squashed version of changes in the side branch,
which is available here: please refer to
https://github.com/bertmaher/pytorch

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Pull changes from bertmaher/pytorch_fusion."


This PR (hopefully the last one of such kind) is merging changes from a
side branch where tensor expessions based fuser work has been done so
far. This PR is is a squashed version of changes in the side branch,
which is available here: https://github.com/bertmaher/pytorch

Differential Revision: [D20478208](https://our.internmc.facebook.com/intern/diff/D20478208)

[ghstack-poisoned]

resistor commented on test/cpp/tensorexpr/test_aten.cpp:
```diff
@@ -363,7 +363,7 @@ void testATenaddcmulFloat() {
     EXPECT_EQ(c_v(i), 3 * i + 2) << "index: " << i;
     EXPECT_EQ(d_v(i), 5 * i + 3) << "index: " << i;
-    EXPECT_EQ(e_v(i), a_v(i) + b_v(i) * c_v(i) * d_v(i)) << "index: " << i;
+    EXPECT_FLOAT_EQ(e_v(i), a_v(i) + b_v(i) * c_v(i) * d_v(i)) << "index: " << i;
```
We should probably be using `EXPECT_FLOAT_EQ` in a lot more places.

Commit:
Update on "[TensorExpr] Pull changes from bertmaher/pytorch_fusion."


This PR (hopefully the last one of such kind) is merging changes from a
side branch where tensor expessions based fuser work has been done so
far. This PR is is a squashed version of changes in the side branch,
which is available here: https://github.com/bertmaher/pytorch

Differential Revision: [D20478208](https://our.internmc.facebook.com/intern/diff/D20478208)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Pull changes from bertmaher/pytorch_fusion."


This PR (hopefully the last one of such kind) is merging changes from a
side branch where tensor expessions based fuser work has been done so
far. This PR is is a squashed version of changes in the side branch,
which is available here: https://github.com/bertmaher/pytorch

Differential Revision: [D20478208](https://our.internmc.facebook.com/intern/diff/D20478208)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Pull changes from bertmaher/pytorch_fusion."


This PR (hopefully the last one of such kind) is merging changes from a
side branch where tensor expessions based fuser work has been done so
far. This PR is is a squashed version of changes in the side branch,
which is available here: https://github.com/bertmaher/pytorch

Differential Revision: [D20478208](https://our.internmc.facebook.com/intern/diff/D20478208)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Pull changes from bertmaher/pytorch_fusion."


This PR (hopefully the last one of such kind) is merging changes from a
side branch where tensor expessions based fuser work has been done so
far. This PR is is a squashed version of changes in the side branch,
which is available here: https://github.com/bertmaher/pytorch

Differential Revision: [D20478208](https://our.internmc.facebook.com/intern/diff/D20478208)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Pull changes from bertmaher/pytorch_fusion."


This PR (hopefully the last one of such kind) is merging changes from a
side branch where tensor expessions based fuser work has been done so
far. This PR is is a squashed version of changes in the side branch,
which is available here: https://github.com/bertmaher/pytorch

Differential Revision: [D20478208](https://our.internmc.facebook.com/intern/diff/D20478208)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Pull changes from bertmaher/pytorch_fusion."


This PR (hopefully the last one of such kind) is merging changes from a
side branch where tensor expessions based fuser work has been done so
far. This PR is is a squashed version of changes in the side branch,
which is available here: https://github.com/bertmaher/pytorch

Differential Revision: [D20478208](https://our.internmc.facebook.com/intern/diff/D20478208)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Pull changes from bertmaher/pytorch_fusion."


This PR (hopefully the last one of such kind) is merging changes from a
side branch where tensor expessions based fuser work has been done so
far. This PR is is a squashed version of changes in the side branch,
which is available here: https://github.com/bertmaher/pytorch

Differential Revision: [D20478208](https://our.internmc.facebook.com/intern/diff/D20478208)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Pull changes from bertmaher/pytorch_fusion."


This PR (hopefully the last one of such kind) is merging changes from a
side branch where tensor expessions based fuser work has been done so
far. This PR is is a squashed version of changes in the side branch,
which is available here: https://github.com/bertmaher/pytorch

Differential Revision: [D20478208](https://our.internmc.facebook.com/intern/diff/D20478208)

[ghstack-poisoned]

Krovatkin reviewed (approved):
:shipit:
